### PR TITLE
Fix legacy task list position calculation

### PR DESCRIPTION
**Pull request by nicolleromero**
This PR fixes a bug where the combination of a tasklist + legacy task list in the side panel would cause a miscalculation of the checkbox position and result in the wrong checkbox being updated. Preventing checkboxes within code blocks from being included in the position calculation prevents this bug and other potential bugs (e.g., where a user includes the `- [ ]` checkbox notation within a code block).

Note: this fix was tested and works in both the previous and new issue viewer in Memex. 

- Closes https://github.com/github/issues/issues/6850

#### Before

https://github.com/primer/react/assets/25596676/76c5c8e7-639a-4049-855d-a7f66845589b

#### After

https://github.com/primer/react/assets/25596676/91521149-dcd7-4a7f-b979-286bf705e02b

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Review by JelloBagel — APPROVED** [review]
Thanks for fixing this bug! ✨ 

Looks good to me. Should we add a test for this?

File: src/drafts/MarkdownViewer/_useListInteraction.ts [hunk 1]
```diff
@@ -4,6 +4,8 @@ import {ListItem, listItemToString, parseListItem} from '../MarkdownEditor/_useL
 
 type TaskListItem = ListItem & {taskBox: '[ ]' | '[x]'}
 
+const isCodeBlockDelimiter = (line: string) => line.trimStart().startsWith('```')
```
Comment (iansan5653) [inline src/drafts/MarkdownViewer/_useListInteraction.ts:7]: The [spec](https://github.github.com/gfm/#fenced-code-blocks) is a little looser than this, so there will be some types of code blocks that don't get caught.

They can be delimited by tildes instead of backticks:

```md
~~~
example
~~~
```

They can also have more than three backticks:

`````md
````
example
````
`````

In this case, you'd need to keep track of how long the starting delimiter is so that you can ensure the ending delimiter is the same length.

Comment (nicolleromero) [reply src/drafts/MarkdownViewer/_useListInteraction.ts:7]: Great point! I think for this specific bug fix (preventing the buggy interaction between legacy task lists and tasklists), the backtick delimiter does the trick, but I'm open to making the fix more robust (checking for tildes) to cover more edge cases.

**Comment by nicolleromero** [comment]
> Thanks for fixing this bug! ✨
> 
> Looks good to me. Should we add a test for this?

@JelloBagel Test added. :)

**Comment by joshblack** [comment]
FYI @siddharthkp for release next week this would be great to get in!

File: src/drafts/MarkdownViewer/_useListInteraction.ts [hunk 2]
```diff
@@ -4,6 +4,17 @@ import {ListItem, listItemToString, parseListItem} from '../MarkdownEditor/_useL
 
 type TaskListItem = ListItem & {taskBox: '[ ]' | '[x]'}
 
+// Make check for code fences more robust per spec: https://github.github.com/gfm/#fenced-code-blocks
+const parseCodeFenceBegin = (line: string) => {
+  const match = line.match(/^ {0,3}(`{3,}|~{3,})[^`]*$/)
```
Comment (iansan5653) [inline src/drafts/MarkdownViewer/_useListInteraction.ts:9]:
```suggestion
  const match = line.match(/^ *(`{3,}|~{3,})[^`]*$/)
```

Looks like the spec says that an opening fence can be indented an indefinite number of spaces, but the closing fence can only be indented up to 3.

File: src/drafts/MarkdownViewer/_useListInteraction.ts [hunk 3]
```diff
@@ -4,6 +4,17 @@ import {ListItem, listItemToString, parseListItem} from '../MarkdownEditor/_useL
...
+  return match ? match[1] : null
+}
+
+const isCodeFenceEnd = (line: string, fence: string) => {
+  const match = line.match(new RegExp(`^ {0,3}${fence}${fence[0]}* *$`))
```
Comment (iansan5653) [inline src/drafts/MarkdownViewer/_useListInteraction.ts:14]:
```suggestion
  const match = line.match(new RegExp(`^ {0,3}${fence} *$`))
```
I think the closing fence needs to have exactly the same number of delimiters as the opening fence. For example, this is valid:

`````md
```one
````
we are still in block `one` here
```
`````

File: src/drafts/MarkdownViewer/_useListInteraction.ts [hunk 4]
```diff
@@ -4,6 +4,17 @@ import {ListItem, listItemToString, parseListItem} from '../MarkdownEditor/_useL
...
+  return match !== null
```
Comment (iansan5653) [inline src/drafts/MarkdownViewer/_useListInteraction.ts:15]: Nit: we could return `regex.test(line)` instead of `line.match(regex) !== null`. I think there might be a small performance benefit to that.

File: src/drafts/MarkdownViewer/_useListInteraction.ts (hunk 2 above)
Comment (nicolleromero) [reply src/drafts/MarkdownViewer/_useListInteraction.ts:9]: The spec states that both an opening and closing fence can be indented no more than three spaces:

> A [fenced code block](https://github.github.com/gfm/#fenced-code-block) begins with a code fence, indented no more than three spaces.

> The closing code fence may be indented up to three spaces

I also tested and confirmed this is the case.

Comment (iansan5653) [reply src/drafts/MarkdownViewer/_useListInteraction.ts:9]: Huh, the spec has an example with more than three: https://github.github.com/gfm/#example-104

But I also tested and confirmed, this doesn't work. So maybe this is a problem with our own pipeline. Guess this is all good then :+1:

Comment (iansan5653) [reply src/drafts/MarkdownViewer/_useListInteraction.ts:9]: Side note, maybe we also should filter out lines starting with more than 3 spaces?

Ie, this is not a checklist item, but it's also not in a code block:

    - [ ] example

This came from this markdown:

```md
    - [ ] example
```

File: src/drafts/MarkdownViewer/_useListInteraction.ts (hunk 4 above)
Comment (nicolleromero) [reply src/drafts/MarkdownViewer/_useListInteraction.ts]: Updated :)

File: src/drafts/MarkdownViewer/_useListInteraction.ts (hunk 3 above)
Comment (iansan5653) [reply src/drafts/MarkdownViewer/_useListInteraction.ts]: Oh wow, you are totally right 👍

**Review by iansan5653 — APPROVED** [review]

**Review by siddharthkp — APPROVED** [review]
